### PR TITLE
feat(host-contracts): introduce the extraData and decryptionProof fields at KMSVerifier

### PR DESCRIPTION
**Pull request by isaacdecoded**
Note that this PR not only introduces the `extraData` field at the `KMSVerifier` but also refactors its `verifyDecryptionEIP712KMSSignatures` method to receive a `bytes decryptionProof` argument containing `number of signers + KMS signatures + extra data`. This refactor is based on the changes requested in the [Add ERC-7995 PR](https://github.com/ethereum/ERCs/pull/1143/files#diff-05bcded41894aaf6a46d15840fd9d378934b820bda8e59c22b0dcfec9fe667a2R342)

Closes https://github.com/zama-ai/fhevm-internal/issues/298

### CODE REVIEW

File: host-contracts/contracts/KMSVerifier.sol [hunk 1]
```diff
@@ -21,6 +21,12 @@ contract KMSVerifier is UUPSUpgradeableEmptyProxy, Ownable2StepUpgradeable, EIP7
     /// @param invalidSigner Address of the invalid signer.
     error KMSInvalidSigner(address invalidSigner);
 
+    /// @notice Returned if the deserializing of the decryption proof fails.
+    error DeserializingDecryptionProofFail();
```
Comment (melanciani) [inline host-contracts/contracts/KMSVerifier.sol:25]: ~~maybe `DeserializingDecryptionProofFailed` , like we use `KMSSignatureThresholdNotReached` below~~

just saw that this the same error as in InputVerifier, all good then

File: host-contracts/contracts/KMSVerifier.sol [hunk 2]
```diff
@@ -169,18 +177,51 @@ contract KMSVerifier is UUPSUpgradeableEmptyProxy, Ownable2StepUpgradeable, EIP7
...
+
+        /// @dev The decryptionProof is the numSigners + kmsSignatures + extraData (1 + 65*numSigners + extraData bytes)
+        uint256 numSigners = uint256(uint8(decryptionProof[0]));
+
+        /// @dev Check that the decryptionProof is long enough to contain at least the numSigners + kmsSignatures
+        if (decryptionProof.length < 1 + 65 * numSigners) {
+            revert DeserializingDecryptionProofFail();
+        }
+
+        bytes[] memory signatures = new bytes[](numSigners);
```
Comment (melanciani) [inline host-contracts/contracts/KMSVerifier.sol:203]: maybe create an issue to factorize this function into a common abstract contract, as it's used in InputVerifier as well ! imo it can be generalized with an offset as input

File: host-contracts/contracts/KMSVerifier.sol [hunk 3]
```diff
@@ -169,18 +177,51 @@ contract KMSVerifier is UUPSUpgradeableEmptyProxy, Ownable2StepUpgradeable, EIP7
...
+                signatures[j][i] = decryptionProof[1 + 65 * j + i];
+            }
+        }
+
+        /// @dev The extraData is the rest of the decryptionProof bytes after the numSigners + signatures.
+        uint256 signaturesSize = uint256(numSigners) * 65;
+        uint256 extraDataOffset = 1 + signaturesSize;
+        uint256 extraDataSize = decryptionProof.length - extraDataOffset;
+
+        bytes memory extraData = new bytes(extraDataSize);
```
Comment (melanciani) [inline host-contracts/contracts/KMSVerifier.sol:213]: same as in the InputVerifier PR, I think you can use slices right ?

File: host-contracts/contracts/KMSVerifier.sol (hunk 2 above)
Comment (melanciani) [reply host-contracts/contracts/KMSVerifier.sol:203]: I wonder if we could use slices here to simplify a bit the logic

File: host-contracts/contracts/KMSVerifier.sol [hunk 4]
```diff
@@ -169,18 +177,51 @@ contract KMSVerifier is UUPSUpgradeableEmptyProxy, Ownable2StepUpgradeable, EIP7
...
+        for (uint256 j = 0; j < numSigners; j++) {
+            signatures[j] = new bytes(65);
+            for (uint256 i = 0; i < 65; i++) {
+                signatures[j][i] = decryptionProof[1 + 65 * j + i];
+            }
+        }
+
+        /// @dev The extraData is the rest of the decryptionProof bytes after the numSigners + signatures.
+        uint256 signaturesSize = uint256(numSigners) * 65;
+        uint256 extraDataOffset = 1 + signaturesSize;
```
Comment (melanciani) [inline host-contracts/contracts/KMSVerifier.sol]: same as in the InputVerifier PR, this offset has already been computed in line 196, I think you can factorize a bit

File: host-contracts/test/kmsVerifier/kmsVerifier.t.sol [hunk 5]
```diff
@@ -379,13 +381,21 @@ contract KMSVerifierTest is Test {
         bytes32[] memory handlesList = _generateMockHandlesList(3);
 
         bytes memory decryptedResult = abi.encodePacked(keccak256("test"), keccak256("test"), keccak256("test"));
-        bytes[] memory signatures = new bytes[](2);
+        bytes memory extraData = abi.encodePacked(uint8(0));
```
Comment (RomanBredehoft) [inline host-contracts/test/kmsVerifier/kmsVerifier.t.sol:384]: could you rename extraData to `decryptionProof` to make name more coherent ? since this is what we provide here right ?

File: host-contracts/test/kmsVerifier/kmsVerifier.t.sol [hunk 6]
```diff
@@ -396,19 +406,27 @@ contract KMSVerifierTest is Test {
...
+        );
+
         vm.expectPartialRevert(KMSVerifier.KMSInvalidSigner.selector);
-        kmsVerifier.verifyDecryptionEIP712KMSSignatures(handlesList, decryptedResult, signatures);
+        kmsVerifier.verifyDecryptionEIP712KMSSignatures(handlesList, decryptedResult, decryptionProof);
     }
 
     /**
-     * @dev Tests that the verification of EIP-712 KMS signatures fails as expected when no signer is added.
+     * @dev Tests that the verification of EIP-712 KMS signatures fails as expected when invalid signature is added.
```
Comment (melanciani) [inline host-contracts/test/kmsVerifier/kmsVerifier.t.sol]: not sure about this comment change

**Review by melanciani — CHANGES_REQUESTED** [review]
thanks, a few changes to request !

File: host-contracts/contracts/KMSVerifier.sol (hunk 2 above)
Comment (isaacdecoded) [reply host-contracts/contracts/KMSVerifier.sol:203]: Tracking issue [here](https://github.com/zama-ai/fhevm-internal/issues/323). Thanks! ✊

File: host-contracts/contracts/KMSVerifier.sol (hunk 3 above)
Comment (isaacdecoded) [reply host-contracts/contracts/KMSVerifier.sol:213]: In the end, we figured out that atm using array slices is for `calldata` arrays only. Clarified also [here](https://github.com/zama-ai/fhevm/pull/666#discussion_r2252091846).

File: host-contracts/test/kmsVerifier/kmsVerifier.t.sol (hunk 6 above)
Comment (isaacdecoded) [reply host-contracts/test/kmsVerifier/kmsVerifier.t.sol]: You're right, "no signer is added" is more accurate since the test is signing with a null signer. Restored [here](https://github.com/zama-ai/fhevm/pull/672/commits/fcb107331653b9464e0d077d07d4044eb9fc5a1b#diff-b6b302406e03e06335a181b8e5d5566a843461272838e5a7bfdd434ee39abc86R429).

File: host-contracts/contracts/KMSVerifier.sol (hunk 4 above)
Comment (isaacdecoded) [reply host-contracts/contracts/KMSVerifier.sol]: Refactor done [here](https://github.com/zama-ai/fhevm/pull/672/commits/fcb107331653b9464e0d077d07d4044eb9fc5a1b#diff-cf8afd41b5bb584dcf568def88aa23b92cf6a33cd35d91a3e892df242566f04aR196). Thanks again for the improvement ✊

**Review by melanciani — APPROVED** [review]
looks good to me, thanks !!
